File: src/main/java/com/dingdongdeng/coinautotrading/common/slack/SlackSender.java
package com.dingdongdeng.coinautotrading.common.slack;

import java.util.HashMap;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

@Slf4j
@RequiredArgsConstructor
@Component
public class SlackSender {

  @Value("${spring.profiles.active:}")
  private String profile;

  @Value("${slack.webHookUrl:}")
  private String webHookUrl;

  private final WebClient slackWebClient;

  public void send(Throwable throwable) {
    send("occured error ::: ", throwable);
  }

  public void send(String message, Throwable throwable) {
    if (!profile.equalsIgnoreCase("release")) {
      return;
    }
    Map<String, String> body = new HashMap<>();
    body.put("username", "coinautotrading");
    body.put("text", message + "\n" + throwable.getMessage());
    try {
      slackWebClient
          .post()
          .uri(webHookUrl)
          .bodyValue(body)
          .retrieve()
          .bodyToMono(Void.class)
          .retry(2)
          .subscribe(res -> log.info("slack sended"), e -> log.error(e.getMessage(), e));
    } catch (Exception e) {
      log.error("Failed to send slack message. error : {}, message : {}", e, body);
    }
  }
}

File: app-service/src/main/kotlin/com/dingdongdeng/autotrading/infra/client/slack/SlackSender.kt
package com.dingdongdeng.autotrading.infra.client.slack

import com.dingdongdeng.autotrading.infra.common.log.Slf4j.Companion.log
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.client.RestClient

@Component
class SlackSender(
    @Value("\${spring.profiles.active:}")
    private val profile: String,
    private val slackRestClient: RestClient,
) {
    fun send(throwable: Throwable) {
        send("occured error ::: ", throwable)
    }

    fun send(message: String, throwable: Throwable) {
        if (profile != "prod") {
            return
        }
        val body: MutableMap<String, String> = HashMap()
        body["username"] = "autotrading"
        body["text"] = """
             $message
             ${throwable.message}
             """.trimIndent()
        try {
            slackRestClient.post()
                .body(body)
                .retrieve()
                .body(Void::class.java)
        } catch (e: Exception) {
            log.error("Failed to send slack message. error : {}, message : {}", e, body)
        }
    }
}
